src/extblkdev/vdo/ExtBlkDevVdo.cc

```diff
@@ -0,0 +1,141 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * (C) Copyright IBM Corporation 2022
+ * Author: Martin Ohmacht <mohmacht@us.ibm.com>
+ *
+ * Based on the file ceph/src/common/blkdev.cc
+ * Copyright (c) 2015 Hewlett-Packard Development Company, L.P.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+
+#include "ExtBlkDevVdo.h"
+#include "common/blkdev.h"
+#include "include/stringify.h"
+#include <errno.h>
+
+int ExtBlkDevVdo::_get_vdo_stats_handle(const std::string& devname)
+{
+  int rc = -ENOENT;
+
+  // we need to go from the raw devname (e.g., dm-4) to the VDO volume name.
+  // currently the best way seems to be to look at /dev/mapper/* ...
+  std::string expect = std::string("../") + devname;  // expected symlink target
+  DIR *dir = ::opendir("/dev/mapper");
+  if (!dir) {
+    return -errno;
+  }
+  struct dirent *de = nullptr;
+  while ((de = ::readdir(dir))) {
+    if (de->d_name[0] == '.')
+      continue;
+    char fn[4096], target[4096];
+    snprintf(fn, sizeof(fn), "/dev/mapper/%s", de->d_name);
+    int r = readlink(fn, target, sizeof(target));
+    if (r < 0 || r >= (int)sizeof(target))
+      continue;
+    target[r] = 0;
+    if (expect == target) {
+      snprintf(fn, sizeof(fn), "/sys/kvdo/%s/statistics", de->d_name);
+      int vdo_fd = ::open(fn, O_RDONLY|O_CLOEXEC); //DIRECTORY);
+      if (vdo_fd >= 0) {
+	name = de->d_name;
+	vdo_dir_fd=vdo_fd;
+	rc=0;
+	break;
+      }
+    }
+  }
+  closedir(dir);
+  return rc;
+}
+
+int ExtBlkDevVdo::get_vdo_stats_handle()
+{
+  std::set<std::string> devs = { logdevname };
+  while (!devs.empty()) {
+    std::string dev = *devs.begin();
+    devs.erase(devs.begin());
+    int rc = _get_vdo_stats_handle(dev);
+    if (rc == 0) {
+      // yay, it's vdo
```

Comment:
+1 for comments like these

Reply:
Ditto, I can not take the credit, the comment is copied verbatim with the rest from the old VDO code.